// @microsoft/mgt-components/dist/es6/components/mgt-planner/task-sources.js
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/* eslint-disable max-classes-per-file */
import { BetaGraph } from '@microsoft/mgt-element';
import { addPlannerTask, assignPeopleToPlannerTask, getAllMyPlannerPlans, getBucketsForPlannerPlan, getPlansForGroup, getSinglePlannerPlan, getTasksForPlannerBucket, removePlannerTask, setPlannerTaskComplete, setPlannerTaskIncomplete } from './graph.planner';
/**
 * async method to get user details
 *
 * @class TaskSourceBase
 */
class TaskSourceBase {
    constructor(graph) {
        // Use an instance of BetaGraph since we know we need to call beta apis.
        this.graph = BetaGraph.fromGraph(graph);
    }
}
/**
 * Create Planner
 *
 * @export
 * @class PlannerTaskSource
 * @extends {TaskSourceBase}
 * @implements {ITaskSource}
 */
export class PlannerTaskSource extends TaskSourceBase {
    /**
     * returns promise with all of users plans
     *
     * @returns {Promise<ITaskGroup[]>}
     * @memberof PlannerTaskSource
     */
    getTaskGroups() {
        return __awaiter(this, void 0, void 0, function* () {
            const plans = yield getAllMyPlannerPlans(this.graph);
            return plans.map(plan => { var _a; return ({ id: plan.id, title: plan.title, containerId: (_a = plan === null || plan === void 0 ? void 0 : plan.container) === null || _a === void 0 ? void 0 : _a.containerId }); });
        });
    }
    /**
     * returns promise with all of plans for group id
     *
     * @param {string} id
     * @returns {Promise<ITaskGroup[]>}
     * @memberof PlannerTaskSource
     */
    getTaskGroupsForGroup(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const plans = yield getPlansForGroup(this.graph, id);
            return plans.map(plan => ({ id: plan.id, title: plan.title }));
        });
    }
    /**
     * returns promise single TaskGroup or plan from plan.id
     *
     * @param {string} id
     * @returns {Promise<ITaskGroup>}
     * @memberof PlannerTaskSource
     */
    getTaskGroup(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const plan = yield getSinglePlannerPlan(this.graph, id);
            return { id: plan.id, title: plan.title, _raw: plan };
        });
    }
    /**
     * returns promise with Bucket for a plan from bucket.id
     *
     * @param {string} id
     * @returns {Promise<ITaskFolder[]>}
     * @memberof PlannerTaskSource
     */
    getTaskFoldersForTaskGroup(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const buckets = yield getBucketsForPlannerPlan(this.graph, id);
            return buckets.map(bucket => ({
                _raw: bucket,
                id: bucket.id,
                name: bucket.name,
                parentId: bucket.planId
            }));
        });
    }
    /**
     * get all task from a Bucket given task id
     *
     * @param {string} id
     * @returns {Promise<ITask[]>}
     * @memberof PlannerTaskSource
     */
    getTasksForTaskFolder(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const tasks = yield getTasksForPlannerBucket(this.graph, id);
            return tasks.map(task => ({
                _raw: task,
                assignments: task.assignments,
                completed: task.percentComplete === 100,
                dueDate: task.dueDateTime && new Date(task.dueDateTime),
                eTag: task['@odata.etag'],
                id: task.id,
                immediateParentId: task.bucketId,
                name: task.title,
                topParentId: task.planId
            }));
        });
    }
    /**
     * set task in planner to complete state by id
     *
     * @param {ITask} task
     * @returns {Promise<any>}
     * @memberof PlannerTaskSource
     */
    setTaskComplete(task) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield setPlannerTaskComplete(this.graph, task);
        });
    }
    /**
     * set task in planner to incomplete state by id
     *
     * @param {ITask} task
     * @returns {Promise<any>}
     * @memberof PlannerTaskSource
     */
    setTaskIncomplete(task) {
        return __awaiter(this, void 0, void 0, function* () {
            return setPlannerTaskIncomplete(this.graph, task);
        });
    }
    /**
     * add new task to bucket
     *
     * @param {ITask} newTask
     * @returns {Promise<any>}
     * @memberof PlannerTaskSource
     */
    addTask(newTask) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            return yield addPlannerTask(this.graph, {
                assignments: newTask.assignments,
                bucketId: newTask.immediateParentId,
                dueDateTime: (_a = newTask.dueDate) === null || _a === void 0 ? void 0 : _a.toISOString(),
                planId: newTask.topParentId,
                title: newTask.name
            });
        });
    }
    /**
     * Assigns people to task
     *
     * @param {ITask} task
     * @param {PlannerAssignments} people
     * @returns {Promise<any>}
     * @memberof PlannerTaskSource
     */
    assignPeopleToTask(task, people) {
        return __awaiter(this, void 0, void 0, function* () {
            return assignPeopleToPlannerTask(this.graph, task, people);
        });
    }
    /**
     * remove task from bucket
     *
     * @param {ITask} task
     * @returns {Promise<any>}
     * @memberof PlannerTaskSource
     */
    removeTask(task) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield removePlannerTask(this.graph, task);
        });
    }
    /**
     * assigns task to the signed in user
     *
     * @param {ITask} task
     * @param {string} myId
     * @returns {boolean}
     * @memberof PlannerTaskSource
     */
    isAssignedToMe(task, myId) {
        const keys = Object.keys(task.assignments);
        return keys.includes(myId);
    }
}
